### PR TITLE
DEV-2518: add tax_id field to revenue program response

### PR DESCRIPTION
**Pull request by leo-dcfa**
Please fill out each section even if it's just with "N/A"

#### Plan? (if this draft/incomplete indicate what you intend to do and how)

This PR has another stacked PR as part of it - Please merge #939 before merging the current PR to `main` 

#### What's this PR do?

- Add `tax_id` to revenue program response from rev-engine
- Makes `tax_id` available in Django admin
- Added new permissions `IsRpAdmin`, `IsOrgAdmin`, `IsGetRequest` and `IsPatchRequest`
- Added patch method to allow patching the tax_id in a revenue program
- Modified Revenue Program view to return responses under following conditions:

1. User is authenticated and
2. User is a super user OR
3. User is an org admin and it's a patch or get request OR
4. User is a rp admin and it's a get request

#### Why are we doing this? How does it help us?

As part of [DEV-3067], we are making org settings modifiable by the Org Admin. [DEV-2518] is about making the tax_id available for edit for the Org Admin and visible for the RP Admin

#### How should this be manually tested? Please include detailed step-by-step instructions.

Please note the test below below covers patching and retrieving a revenue program with a super user. I've added multiple tests in the unit tests to cover different types of permissions. The test below verifies patching and retrieving revenue program + asserting that tax_id is included in response

- Start the app with `make run-dev`
- Navigate to Django Admin and open any revenue program
- Copy the id for the revenue program
- Run ` curl -c cookiejar -v -H "Content-Type: application/json" -H "Accept: application/json" -X POST -d '{"email":"{your username}", "password": "{your password}"}' http://localhost:8000/api/v1/token/`
- As part of the response from the above, a CSRF token is returned - you will use the token in
- Run ` curl -b cookiejar -v -H "Content-Type: application/json" -H "Accept: application/json" -H "X-CSRFToken: {CSRF TOKEN from response above}"  -X PATCH "http://localhost:8000/api/v1/revenue-programs/{revenue program id from above}/" --data '{"tax_id": "223456789"}'\ | jq .` and assert the `tax_id` was updated
- Run ` curl -b cookiejar -v -H "Content-Type: application/json" -H "Accept: application/json" -H "X-CSRFToken: {CSRF TOKEN from response above}" -X GET "http://localhost:8000/api/v1/revenue-programs/18/" \ |jq .`
- Assert that `tax_id` was included in the response

#### Did this PR make changes to the user interface that may require changes to the user-facing docs?

N/A

#### Have automated unit tests been added? If not, why?

Yes

#### How should this change be communicated to end users?

N/A

#### Are there any smells or added technical debt to note?

N/A

#### Has this been documented? If so, where?

N/A

#### What are the relevant tickets? Add a link to any relevant ones.

[DEV-2518]

#### Are there next steps? If so, what? Have tickets been opened for them? List next-step tickets here:

[DEV-2518]: https://news-revenue-hub.atlassian.net/browse/DEV-2518?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[DEV-3067]: https://news-revenue-hub.atlassian.net/browse/DEV-3067?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/organizations/tests/test_org_views.py [hunk 1]
```diff
@@ -100,7 +100,7 @@ def test_superuser_can_list_rps(self):
         return self.assert_superuser_can_list(self.list_url, expected_count, results_are_flat=True)
 
     def test_other_cannot_access_resource(self):
-        non_superusers = [self.hub_user, self.org_user, self.rp_user, self.contributor_user]
+        non_superusers = [self.hub_user, self.contributor_user]
```
Comment (benjaminEwhite) [inline apps/organizations/tests/test_org_views.py:103]: There are some critical paths where there are missing tests, as well as data being leaked between orgs/RPs. Specifically, we need tests per method/endpoint that show that a user of a given type (in this case org user) who has access to the endpoints as such nevertheless does not other people's data in responses.

I added the following tests locally, which fail. Could probably be cleaned up, but this is what I'm talking about:


```
def test_cannot_retrieve_someone_elses_rp(self):
        other_org_user = create_test_user(
            role_assignment_data={"role_type": Roles.ORG_ADMIN, "organization": self.org2}
        )
        assert other_org_user.roleassignment.organization != self.org1
        self.assert_user_cannot_get(self.detail_url, other_org_user)

 def test_other_orgs_rps_not_appear_in_my_list_of_rps(self):
        other_org_user = create_test_user(
            role_assignment_data={"role_type": Roles.ORG_ADMIN, "organization": self.org2}
        )
        assert other_org_user.roleassignment.organization != self.org1
        assert (expected_rp_count := other_org_user.roleassignment.organization.revenueprogram_set.count()) == 1

        response = self.assert_user_can_get(self.list_url, other_org_user)
        assert len(response.json()) == expected_rp_count
        # assert that all returned have ids owned by requesting user etc.
```

Comment (benjaminEwhite) [reply apps/organizations/tests/test_org_views.py:103]: These kinds of cases are covered extensively in the unit tests for contributions views, if you want to see how I handled testing (and implementation) there.

File: apps/organizations/views.py [hunk 2]
```diff
@@ -46,11 +53,27 @@ def get_queryset(self):
 class RevenueProgramViewSet(viewsets.ReadOnlyModelViewSet):
```
Comment (benjaminEwhite) [inline apps/organizations/views.py:53]: Now that we're adding `patch` to this viewset, might be better to not inherit from ReadOnlyModelViewset. As currently implemented, the `patch` method will run and works, but  when you see that this is ReadOnlyModelViewSet you don't expect it to have `patch`. I'd prefer to see a combination of viewset mixins, or barring that, perhaps a code comment here explaining why there's a patch on a readonly model view set.

File: apps/organizations/views.py [hunk 3]
```diff
@@ -46,11 +53,27 @@ def get_queryset(self):
...
+    ]
     serializer_class = serializers.RevenueProgramSerializer
     pagination_class = None
 
+    def patch(self, request, pk):
+        revenue_program = get_object_or_404(RevenueProgram, pk=pk)
+        serializer = RevenueProgramPatchSerializer(revenue_program, data=request.data, partial=True)
+        serializer.is_valid()
+        if serializer.errors:
+            errors = {**serializer.errors}
```
Comment (benjaminEwhite) [inline apps/organizations/views.py:70]: Is this casting something dict like to dict, or to avoid mutation, or what?

**Review by benjaminEwhite — CHANGES_REQUESTED** [review]
Left a few comments. Biggest thing is we need to prevent org users from having access to other org's data, and that will require some refactoring to achieve.

File: apps/api/permissions.py [hunk 4]
```diff
@@ -45,6 +45,46 @@ def has_permission(self, request, view):
...
+                role_assignment.role_type == Roles.ORG_ADMIN,
+            ]
+        )
+
+
+class IsRpAdmin(permissions.BasePermission):
+    """Org Admin permission
+
+    If the user making the request is an org admin, grant permission.
+    """
```
[the remark this answers is not in the document]
Comment (leo-dcfa) [reply apps/api/permissions.py:67]: good catch, thanks

File: apps/organizations/views.py (hunk 3 above)
Comment (leo-dcfa) [reply apps/organizations/views.py:70]: ah thanks I copied a template piece of code to get started and didn't realise that it wasn't needed

File: apps/organizations/views.py (hunk 2 above)
Comment (leo-dcfa) [reply apps/organizations/views.py:53]: thanks, good point

File: apps/organizations/tests/test_org_views.py (hunk 1 above)
Comment (leo-dcfa) [reply apps/organizations/tests/test_org_views.py:103]: Thanks for that, code updated

**Review by benjaminEwhite — APPROVED** [review]

**Comment by benjaminEwhite** [comment]
Closing, tearing down, then reopening to trigger new review app build, as part of [DEV-3243 | fix hookdeck webhooks](https://news-revenue-hub.atlassian.net/browse/DEV-3243).

**Comment by benjaminEwhite** [comment]
Closing and reopening (after dealing with merge conflicts) to solve for migration conflict that will break review app deploys.

**Review by benjaminEwhite — APPROVED** [review]

**Review by x110dc — APPROVED** [review]

**Pull request merged**